### PR TITLE
Add a request body size limit that closes over-limit connections

### PR DESCRIPTION
**Pull request by anton-n-petrov**
Addresses #1064.

## PR 1 of 2

This is a split of #1233, following [@ssubbotin's review](https://github.com/CrowCpp/Crow/pull/1233#issuecomment-5482035385) ("Step 1, its own PR: a general request body limit") and [@gittiver's comment](https://github.com/CrowCpp/Crow/pull/1233#pullrequestreview-5069820951) asking for a stream interface rather than a file-only feature. #1233 is closed in favor of this PR and a follow-up.

- **PR 1 (this one):** a general request body size limit — the actual fix for #1064, independent of any file/sink feature.
- **PR 2 (#1235, `feature/request-body-sink`):** a pluggable `BodySink` primitive (`.body_sink(factory)`), with `crow::FileBodySink` as a built-in file-backed sink, built on top of this PR's limit.

## Why a separate PR

`#1233` coupled the size limit to a single opted-in file-storage route, so every other route (including 404/405/redirect paths) stayed unbounded, and the limit itself was enforced *after* the bytes were read rather than at headers-complete. Neither problem needs a file feature to fix, so it is its own PR.

## API

```cpp
app.max_body_size(64ull * 1024 * 1024); // default: UINT64_MAX = unlimited

CROW_ROUTE(app, "/upload")
  .methods(crow::HTTPMethod::Post)
  .max_body_size(8ull * 1024 * 1024)    // per-route override
  ([](const crow::request& req) { ... });
```

## Behaviour

- An advertised `Content-Length` over the limit is rejected in `handle_header()`, before `Expect: 100-continue` is answered and before any body bytes are read.
- Chunked bodies (no advertised length) are counted in `on_body` and close on exceeding the limit; the overflow-safe check (`length > max` before `max - length`) avoids wraparound.
- Either way, the connection lingers instead of closing outright: the write side is shut down (FIN, not RST) once the 413 is sent, and whatever the client keeps sending is drained and discarded, so a client still mid-upload isn't reset before it can read the response.
- The cap applies on every path that can reach `handle_header()`: matched routes, 404, 405/catch-all, and slash-redirects.
- Global after-handlers still run on the 413 (e.g. CORS stamps the response); before-handlers do not, matching the existing early-rejection paths for unmatched routes and auto-OPTIONS.
- `req.body` is never pre-`reserve()`d from the (untrusted) `Content-Length` header — it grows only as bytes actually arrive, whether the configured limit is finite or unlimited.
- Default stays unlimited; existing behaviour for apps that don't opt in is unchanged.

## Testing

`tests/max_body_size_tests.cpp` covers the advertised-length/100-continue interaction, chunked accumulation and the exact-boundary case, 404/405/redirect enforcement, after-handler execution on 413, the HEAD/OPTIONS exemptions, a connection-leak regression, and the unlimited default. Full suite (`ctest`/`unittest`) passes: 1039 assertions, 138 test cases.

### CODE REVIEW

**Comment by ssubbotin** [comment]
Reviewed at `225034f8e` against `e91d0e75c`. I read the diff and the surrounding request path in full, built the branch with `-DCROW_BUILD_TESTS=ON`, and checked the load-bearing claims against a server built from these headers. Items marked **[measured]** were reproduced against that running server.

The branch is healthy: the full suite passes, 1015 assertions in 132 test cases. The "1102 assertions, 143 test cases" in the description is the number for the stacked #1235 branch, so the text should be corrected. The mechanism is the right one and it works as advertised: `Content-Length` above the cap is refused in `handle_header()` before `100 Continue` **[measured]**, the boundary is exact (1024 accepted, 1025 refused) **[measured]**, a per-route override is honoured for a `HEAD` that falls back to the `GET` rule and for a route registered through a Blueprint **[measured]**, the limit is re-resolved on every keep-alive request **[measured]**, the cap is applied before the `Host` check and on 404, 405 and slash-redirect paths **[measured]**, `max_body_size(0)` means "no body allowed" and still accepts `Content-Length: 0` **[measured]**, and curl reports 413 for a 5 MB upload in 10 of 10 runs with and without `Expect` **[measured]**. The chunked arithmetic (`length > max` before `max - length`) is correct.

My overall read is **merge after two fixes**, one of which is the only substantive gap in the design.

## Before merge

### 1. The 413 is sent and then the socket is closed with unread bytes in the receive queue

`handle_header()` returning 1 sets `F_SKIPBODY` (`http_parser_merged.h:1633-1634`), the parser fires `on_message_complete` at once (`:1666-1667`), `handle()` writes the 413 synchronously through `do_write_sync`, `on_message_complete` returns 1, `feed()` returns false, and `do_read()` takes the `error_while_reading` branch: `shutdown_read()` then `close()`. For a client that is still sending, `close()` with unread data makes the kernel send RST.

What that looks like from the client: for a 5 MB upload against a 1 KiB cap, a plain socket client's `send()` fails with `ECONNRESET` in 20 of 20 runs; a `recv()` issued after that failure still returns the 413 on Linux loopback **[measured]**. So the status line does reach the wire (the write is synchronous), and curl, which reads while it sends, is fine. What is not fine: any client library that treats the send failure as fatal never reads the 413 (this is the classic "connection reset by peer" report on uploads), receive stacks that purge buffered data on RST lose it, and on a real network the RST can arrive before the client has drained its receive buffer. nginx has `lingering_close` for exactly this case, and Go's `net/http` has `closeWriteAndWait`. The tests cannot see any of it: the test client writes the whole request with one `asio::write` before reading (`tests/max_body_size_tests.cpp:62-65`), every over-limit body is 100 bytes or fewer, and the 5,000,000-byte case uses `Expect` and sends no body.

The fix is small and bounded, and it does not reintroduce draining in the sense the #1233 review objected to. After the 413/500 has been written: `shutdown_write()` (FIN after the response, never RST), re-arm the existing deadline timer (`timeout_`, 5 s by default, whose expiry already does `shutdown_readwrite()` + `close()`), and loop `async_read_some` into `buffer_` discarding everything until the read returns an error (EOF or reset), then `cancel_deadline_timer()` and `close()`. A client that honours `Connection: close` stops within one round trip; one that does not is cut off by the same timer budget a slow client already gets. `shutdown_write()` exists on the TCP, Unix and SSL adaptors already. The same path should serve the chunked reject and, in #1235, the sink-failure 500.

For the test: the observable that distinguishes lingering close from the current behaviour is on the send side. A client uploading well over the cap must be able to write its whole body without an error *and* read the 413. "413 then EOF" passes today for a concurrently reading client on loopback, so it would not catch a regression.

### 2. Drop the `reserve()` from `Content-Length`

`handle_header()` reserves `content_length` bytes whenever a finite limit is set (`http_connection.h:138-145`). Bounded by the cap, yes, but it is allocated at headers-complete, before a single body byte arrives, per connection. With a route cap of 1 MiB, 300 idle connections each advertising `Content-Length: 1048576` and sending nothing grew the server's virtual size by 300 MiB (resident memory by 3 MiB, since glibc hands out untouched pages) **[measured]**. On a system with strict overcommit, or with an allocator that touches pages, that is a real allocation, and `std::bad_alloc` thrown from inside `on_headers_complete` unwinds through `http_parser_execute` into the worker loop's `catch (std::exception&)` (`http_server.h:205-208`): the connection is dropped without a response. The regression test at `:346-381` cannot detect that on Linux, because the caught exception leaves the worker running and the second client is served regardless. The benefit is one avoided growth sequence per request, which nobody has measured. The second commit already contains the fix for the unlimited case; I would remove the reserve altogether rather than keep it behind the finite-limit condition.

## Non-blocking

- **`app.effective_max_body_size()` and `Router::effective_max_body_size()` are connection internals in public API.** The lookup re-derives the matched rule from `found.method` and `found.rule_index` with four guard clauses that duplicate `Router::handle`'s knowledge of `RULE_SPECIAL_REDIRECT_SLASH`, `catch_all` and `InternalMethodCount`. `handle_initial()` has the rule in hand when it fills `found`; a `const BaseRule* rule{nullptr}` on `routing_handle_result` lets the connection compute the limit with no new app or router method, and #1235 immediately shows where the current pattern leads (three near-identical copies of the lookup). If you prefer to keep the method, make it a friend-visible internal as `response` does for `Connection`.
- **`max_body_size_` + `max_body_size_override_` on `BaseRule`** is a hand-rolled `std::optional<uint64_t>`, and `routing.h` already includes `<optional>`.
- **`HTTPMethod method{}` is `Delete`**, a valid table index, so the sentinel protects nothing; `InternalMethodCount` is the sentinel the four-argument constructor already uses. The NSDMI itself is an improvement: it fixes a pre-existing uninitialised read for results built by `Trie::find`.
- **`HEAD` over the cap sends a body with `Content-Length: 0`.** `handle_initial` sets `res.skip_body` for `HEAD`, the 413 keeps it, `end()` computes a zero length, and `write_header_into_buffer` (`http_response.h:415-416`) then appends the automatic `413 Payload Too Large` body anyway: `Content-Length: 0` followed by 23 bytes **[measured]**. This is pre-existing (`HEAD` on a route that answers an empty 404 does the same on the base commit **[measured]**), so it is a master bug that this PR makes reachable on one more path. The one-line fix is to skip the automatic body when `skip_body` is set; it can ride along or go separately.
- **Docs overstate the 404 coverage.** `docs/guides/app.md` says the cap applies to 404. It applies to every request that reads a body; an unmatched `HEAD` or `OPTIONS` is answered in `handle_url()` before headers complete, never reaches `handle_header()`, and reads no body (`parser_.done()` then close). Say that, and say that only global after-handlers run on the 413 (route-local middleware and before-handlers do not).
- **The chunked reject leaves an accepted prefix in `req.body`**, while the `Content-Length` reject leaves it empty. After-handlers see different requests for the same outcome; `req_.body.clear()` in the reject block makes them consistent.
- **Re-entrancy.** `reject_payload_too_large()` runs `handle()`, the synchronous write and `parser_.clear()` from inside `on_body`, and the `Content-Length` path does the same from inside `on_message_complete`. It works because the vendored parser reads `parser->state` from the struct on every iteration and both callbacks return 1 immediately afterwards, so nothing else is parsed. It is still a landmine: any later code that reads parser fields after `handle()` sees a cleared parser. Having the callbacks only record an outcome and letting the connection act after `feed()` returns pairs naturally with the lingering close above.
- **`Content-Length` and `Transfer-Encoding: chunked` together** are rejected by the vendored parser before `on_headers_complete`, and Crow closes without a response **[measured]**. Pre-existing and unrelated, noted for completeness.
- **Default.** I would keep `UINT64_MAX` in this PR: `websocket_max_payload` defaults to unlimited, and a header-only library fetched from `master` cannot signal a behaviour change at compile time, so a finite default would silently 413 working uploads. The cost is that #1064 stays open for applications that never call `max_body_size()`. Please make that trade explicit: a prominent paragraph in the guide and `app.max_body_size(...)` in the hello-world examples, and consider a finite default for the next major release.
- **Tests worth adding:** the mid-upload client described above (`Content-Length` and chunked); a within-limit request followed by an over-limit one on the same socket; `HEAD` on a `GET` route with an override; a Blueprint route; the chunked exact boundary (the current chunked test is 10 against 8); a per-route limit smaller than the app limit. The hand-rolled `TestClient` is copied verbatim into `body_file_tests.cpp` in #1235; a shared test header would avoid that.

Thanks for splitting this out and for getting the ordering right: refusing before `100 Continue` and closing instead of draining is the correct policy, and the per-route override through `RuleParameterTraits` is where it belongs. The lingering close is the one piece that makes the refusal reliably visible to the client that triggered it.

**Comment by anton-n-petrov** [comment]
Pushed `d24c98755`, `5d0fc079e`, `7a1890f86` in response to this review.

**Before merge**
1. Lingering close — `d24c987` adds `linger_close()`/`do_linger_read()`: `shutdown_write()` (FIN, not RST) after the 413, then drains via `async_read_some` until EOF/error, bounded by the existing deadline timer (`timeout_`, default 5s). Two regression tests: a client writes a 5MB over-limit body and confirms both the write and the 413 read succeed (`d24c987`), and a small-body 413 doesn't leak the connection (`7a1890f`).
2. `reserve()` from `Content-Length` — dropped entirely in `d24c987`, no longer conditioned on a finite cap.

**Non-blocking, also picked up in `d24c987`/`5d0fc07`**
- `req_.body.clear()` on the 413 path, so after-handlers see a consistent (empty) body for both the chunked and Content-Length rejects.
- `write_header_into_buffer` skips the automatic status-text body when `skip_body` is set — fixes the `HEAD` double-body bug.
- `routing_handle_result::method` now defaults to `InternalMethodCount`, not the misleading `Delete`.
- `BaseRule`'s override is now `std::optional<uint64_t>`.
- `app.md` corrected: the cap doesn't reach an unmatched `HEAD`/`OPTIONS`, only global after-handlers run on the 413, and the drain wording.
- New coverage: `max_body_size(0)`, a successful chunked body under the limit (`5d0fc07`); unmatched `HEAD`/`OPTIONS` bypass (`7a1890f`).

**Left open**
- The shared `TestClient` header across `max_body_size_tests.cpp`/`body_file_tests.cpp` — still duplicated.
- `Content-Length` + `Transfer-Encoding: chunked` together — confirmed pre-existing and unrelated to this PR, untouched.
- Kept the unlimited default, per your recommendation.
- The `matched_rule()` consolidation for the rule lookup landed on #1235 (which had three call sites); this branch only has the one `effective_max_body_size` call site, so it wasn't ported back here.

Full suite passes on this branch's current tip: 1030 assertions, 137 test cases.

**Comment by ssubbotin** [comment]
The implementation and tests look good after the latest updates. Please refresh the PR description before merge: it still says over-limit `Content-Length` requests use no draining and that finite limits pre-reserve `req.body`, while the current code drains during the lingering close and removed the `reserve()` path. The PR 2 summary also still mentions the removed `.body_file()` API.

File: include/crow/http_connection.h [hunk 1]
```diff
@@ -408,9 +448,20 @@ namespace crow
...
-                      self->adaptor_.shutdown_read();
-                      self->adaptor_.close();
                       CROW_LOG_DEBUG << self << " from read(1) with description: \"" << http_errno_description(static_cast<http_errno>(self->parser_.http_errno)) << '\"';
+                      if (self->payload_too_large_)
+                      {
+                          // The rejection response is already written; shut down the
+                          // write side and drain whatever the client still sends
+                          // instead of closing on unread bytes, which can RST a peer
+                          // that is still mid-upload before it reads the response.
+                          self->linger_close();
```
Comment (ssubbotin) [inline include/crow/http_connection.h:458]: [P1] `linger_close()` is bypassed when the 413 body takes the streamed response path. With `app.stream_threshold(1)`, `do_write_general()` reaches its `close_connection_` block and calls `shutdown_readwrite()` plus `close()` before control returns here. I reproduced this by sending only headers with `Content-Length: 5000000`, reading the 413, and then sending the body: `asio::write` fails with `Broken pipe`. Please apply the body-error guard already present in #1235 (`close_connection_ && !body_error_status_`) here as `!payload_too_large_`, including the static response path, and add the same regression test.

File: docs/guides/app.md [hunk 2]
```diff
@@ -52,6 +52,25 @@ app.tcp_nodelay(true)
...
+
+The advertised `Content-Length` is checked when headers complete, before `100 Continue`; no body bytes are read at all if it is already over the cap. Chunked bodies, and bodies with neither `Content-Length` nor `Transfer-Encoding` (read until the connection closes), are counted as they arrive instead, since their size isn't known upfront. An over-limit request is answered `413 Payload Too Large`; the body is not stored and the route handler does not run. If the body was still arriving when the cap was hit, the connection lingers rather than closing outright: the write side is shut down once the 413 is sent, and whatever the client keeps sending is drained and discarded, so a client still mid-upload isn't reset before it can read the response. The cap applies to every request that reaches this check, including matched routes, 404, 405, and slash-redirects — but not an unmatched `HEAD` or `OPTIONS` request, which is answered before headers complete and reads no body either way. Only global after-handlers run on the 413; route-local middleware and before-handlers do not.
```
Comment (ssubbotin) [inline docs/guides/app.md:60]: [P2] This framing claim does not match the parser. When a request has neither `Content-Length` nor `Transfer-Encoding`, `s_headers_done` assumes a zero-length body and invokes `message_complete`; it does not enter `s_body_identity_eof`. Please remove the read-until-close statement. Also avoid saying that no body bytes are read at all, since the socket read containing the headers can already contain body bytes; describe that over-limit advertised bodies are not parsed or stored instead.

**Comment by gittiver** [comment]
The failing build is my fault, the code from POC taken over to unit tests does not build on windows. 
COuld you rebase to master, then it should work?
